python/ray/train/examples/pipeline_parallelism_example.py

```diff
@@ -0,0 +1,352 @@
+
+        pe = torch.zeros(max_len, d_model)
+        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
+        div_term = torch.exp(
+            torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model)
+        )
+        pe[:, 0::2] = torch.sin(position * div_term)
+        pe[:, 1::2] = torch.cos(position * div_term)
+        pe = pe.unsqueeze(0).transpose(0, 1)
+        # https://github.com/pytorch/pytorch/issues/68407
+        self.register_parameter("pe", nn.Parameter(pe, requires_grad=False))
+
+    def forward(self, x):
+        x = x + self.pe[: x.size(0), :]
+        return self.dropout(x)
+
+
+class Encoder(nn.Module):
+    def __init__(self, ntoken, ninp, dropout=0.5):
+        super(Encoder, self).__init__()
+        self.pos_encoder = PositionalEncoding(ninp, dropout)
+        self.encoder = nn.Embedding(ntoken, ninp)
+        self.ninp = ninp
+        self.init_weights()
+
+    def init_weights(self):
+        initrange = 0.1
+        self.encoder.weight.data.uniform_(-initrange, initrange)
+
+    def forward(self, src):
+        # Need (S, N) format for encoder.
+        src = src.t()
+        src = self.encoder(src) * math.sqrt(self.ninp)
+        return self.pos_encoder(src)
+
+
+class Decoder(nn.Module):
+    def __init__(self, ntoken, ninp):
+        super(Decoder, self).__init__()
+        self.decoder = nn.Linear(ninp, ntoken)
+        self.init_weights()
+
+    def init_weights(self):
+        initrange = 0.1
+        self.decoder.bias.data.zero_()
+        self.decoder.weight.data.uniform_(-initrange, initrange)
+
+    def forward(self, inp):
+        # Need batch dimension first for output of pipeline.
+        return self.decoder(inp).permute(1, 0, 2)
+
+
+def run_worker(config):
+    num_gpus = config["num_gpus"]  # Num gpus for model parallelism.
+    epochs = config["epochs"]  # The number of epochs.
+
+    world_rank = train.world_rank()
+    local_rank = train.local_rank()
+    world_size = train.world_size()
+
+    def print_with_rank(msg):
+        print("[LOCAL RANK {} | WORLD RANK {}]: {}".format(local_rank, world_rank, msg))
+
+    from torchtext.datasets import WikiText2
+    from torchtext.data.utils import get_tokenizer
+    from torchtext.vocab import build_vocab_from_iterator
+
+    train_iter = WikiText2(split="train")
+    tokenizer = get_tokenizer("basic_english")
+    vocab = build_vocab_from_iterator(map(tokenizer, train_iter), specials=["<unk>"])
+    vocab.set_default_index(vocab["<unk>"])
+
+    def data_process(raw_text_iter):
+        data = [
+            torch.tensor(vocab(tokenizer(item)), dtype=torch.long)
+            for item in raw_text_iter
+        ]
+        return torch.cat(tuple(filter(lambda t: t.numel() > 0, data)))
+
+    train_iter, val_iter, test_iter = WikiText2()
+    train_data = data_process(train_iter)
+    val_data = data_process(val_iter)
+    test_data = data_process(test_iter)
+
+    device = torch.device(num_gpus * local_rank)
```

Comment:
can we use train.torch.get_device() here?

Reply:
Ah this actually doesn't work because the current implementation assumes `device == local_rank`. We need to expose a new API to get the list of devices for the multi-GPU case. I can create an issue to track this!